### PR TITLE
Some threadsafety for Configuration

### PR DESCRIPTION
**Pull request by melchiork**
Configuration.Current.ValueFormatterAssembly and Configuration.Current.ValueFormatterDetectionMode can be accessed in concurrent execution safely. Spotted in #664.

### CODE REVIEW

File: Src/FluentAssertions/Common/Configuration.cs [hunk 1]
```diff
@@ -35,12 +37,17 @@ public ValueFormatterDetectionMode ValueFormatterDetectionMode
         {
             get
             {
+                Monitor.Enter(propertiesAccessLock);
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Common/Configuration.cs]: Why not just use `lock`?

**Comment by jnyrup** [comment]
@melchiork Thanks for having a look as this.

I would prefer using the `lock` statement instead.
Currently if `DetermineFormatterDetectionMode()` throws an exception `Monitor.Exit()` won't be called and causing a deadlock.

On the other hand the `lock` statement is [translated](https://blogs.msdn.microsoft.com/ericlippert/2009/03/06/locks-and-exceptions-do-not-mix/) into the following to ensure that `Monitor.Exit` is called.
```
bool lockWasTaken = false;
var temp = obj;
try { Monitor.Enter(temp, ref lockWasTaken); { body } }
finally { if (lockWasTaken) Monitor.Exit(temp); }
```

Out of pure curiosity, do you have any numbers on how taking an exclusive locks influences running all tests in parallel?

**Comment by melchiork** [comment]
I've changed to use `lock` statement instead of Monitor. I was not aware of the thing that @jnyrup pointed out - thanks!

**Comment by dennisdoomen** [comment]
Does this solve #664?

**Comment by melchiork** [comment]
@dennisdoomen As far as I know.

**Pull request merged**